File: tools/xgm2tool/src/sgdk/xgm2tool/format/XGMSample.java
package sgdk.xgm2tool.format;

import java.io.IOException;
import java.lang.reflect.Field;
import java.util.Arrays;

import com.musicg.fingerprint.FingerprintManager;
import com.musicg.fingerprint.FingerprintSimilarity;
import com.musicg.fingerprint.FingerprintSimilarityComputer;
import com.musicg.wave.Wave;
import com.musicg.wave.WaveHeader;

import sgdk.tool.MathUtil;
import sgdk.xgm2tool.Launcher;
import sgdk.xgm2tool.format.SampleBank.InternalSample;
import sgdk.xgm2tool.tool.Util;

/**
 * XGM Sample have fixed sample rate of 13.3 or 6.65 Khz.<br>
 * Size and Offset are 64 bytes boundary.
 * 
 * @author Stephane
 */
public class XGMSample
{
    final static WaveHeader xgmWaveHeader = new WaveHeader();

    {
        xgmWaveHeader.setBitsPerSample(8);
        xgmWaveHeader.setSampleRate(13300);
        xgmWaveHeader.setChannels(1);
    }

    final static int XGM_FULL_RATE = 13300;
    final static int XGM_HALF_RATE = (XGM_FULL_RATE / 2);

    int id;
    final byte[] data;

    // used only for VGM to XGM conversion
    final boolean halfRate;
    final int originId;
    final int originAddr;

    public XGMSample(int id, byte[] data, boolean halfRate, int originId, int originAddr)
    {
        super();

        this.id = id;
        this.data = data;

        // only used for VGM to XGM conversion
        this.halfRate = halfRate;
        this.originId = originId;
        this.originAddr = originAddr;
    }

    public XGMSample(int id, byte[] data)
    {
        this(id, data, false, -1, 0);
    }

    public XGMSample(int id, byte[] data, int offset, int len)
    {
        this(id, Arrays.copyOfRange(data, offset, offset + len));
    }

    public static XGMSample createFromVGMSample(int id, InternalSample sample) throws IOException
    {
        final boolean half = (sample.rate < 10000);

        return new XGMSample(id, Util.resample(sample.getBank().data, sample.addr, sample.len, sample.rate, half ? XGM_HALF_RATE : XGM_FULL_RATE, 256), half,
                sample.id, sample.addr);
    }

    public int getLength()
    {
        return data.length;
    }

    public double getSimilarityScore(XGMSample originSample)
    {
        final int deltaSize = originSample.data.length - data.length;
        // origin sample is longer ? cannot use a shorter sample to replace it...
        if (deltaSize > 150)
            return 0d;

        final int minSize = Math.min(data.length, originSample.data.length);

        // advanced sample compare using fingerprint
        if (Launcher.sampleAdvancedCompare)
        {
            int minSizePadded = (minSize < 2048) ? (int) MathUtil.nextPow2(minSize) : minSize;

            final FingerprintManager fingerprintManager = new FingerprintManager();

            try
            {
                Field sampleSizePerFrameF = FingerprintManager.class.getDeclaredField("sampleSizePerFrame");
                sampleSizePerFrameF.setAccessible(true);

                if (minSize < 512)
                    sampleSizePerFrameF.set(fingerprintManager, Integer.valueOf(256));
                else if (minSize < 1024)
                    sampleSizePerFrameF.set(fingerprintManager, Integer.valueOf(512));
                else if (minSize < 2048)
                    sampleSizePerFrameF.set(fingerprintManager, Integer.valueOf(1024));
                else if (minSize < 4096)
                    sampleSizePerFrameF.set(fingerprintManager, Integer.valueOf(2048));
                else
                    // default
                    sampleSizePerFrameF.set(fingerprintManager, Integer.valueOf(2048));
            }
            catch (Exception e)
            {
                // pad to 2048
                minSizePadded = Math.max(minSize, 2048);
            }

            final byte[] sample1 = new byte[minSizePadded];
            Arrays.fill(sample1, (byte) 128);
            for (int i = 0; i < minSize; i++)
                sample1[i] = (byte) (data[i] + 128);

            final byte[] sample2 = new byte[minSizePadded];
            Arrays.fill(sample2, (byte) 128);
            for (int i = 0; i < minSize; i++)
                sample2[i] = (byte) (originSample.data[i] + 128);

            final Wave wave1 = new Wave(xgmWaveHeader, sample1);
            final Wave wave2 = new Wave(xgmWaveHeader, sample2);
            final byte[] fp1 = fingerprintManager.extractFingerprint(wave1);
            final byte[] fp2 = fingerprintManager.extractFingerprint(wave2);
            final FingerprintSimilarity similarity = new FingerprintSimilarityComputer(fp2, fp1).getFingerprintsSimilarity();

            if (Launcher.verbose)
                System.out.println("similarity between #" + id + " (" + originAddr + ") and #" + originSample.id + " (" + originSample.originAddr + ") = "
                    + similarity.getSimilarity() + " - score = " + similarity.getScore());

            return similarity.getSimilarity() * similarity.getScore();
        }

        // simple sample compare (exact match)
        return Arrays.equals(Arrays.copyOf(data, minSize), Arrays.copyOf(originSample.data, minSize)) ? 1d : 0d;
    }
}

File: tools/crates/xgm2tool/src/xgm_sample.rs
use crate::sample_bank::InternalSample;
use crate::util;

/// XGM Sample have fixed sample rate of 13.3 or 6.65 Khz.
/// Size and Offset are 64 bytes boundary.
pub const XGM_FULL_RATE: i32 = 13300;
pub const XGM_HALF_RATE: i32 = XGM_FULL_RATE / 2;

#[derive(Clone, Debug)]
pub struct XGMSample {
    pub id: i32,
    pub data: Vec<u8>,
    pub half_rate: bool,
    pub origin_id: i32,
    pub origin_addr: i32,
}

impl XGMSample {
    pub fn new(id: i32, data: Vec<u8>, half_rate: bool, origin_id: i32, origin_addr: i32) -> Self {
        XGMSample { id, data, half_rate, origin_id, origin_addr }
    }

    pub fn new_simple(id: i32, data: Vec<u8>) -> Self {
        Self::new(id, data, false, -1, 0)
    }

    pub fn new_from_slice(id: i32, data: &[u8], offset: usize, len: usize) -> Self {
        Self::new_simple(id, data[offset..offset + len].to_vec())
    }

    pub fn create_from_vgm_sample(id: i32, sample: &InternalSample, bank_data: &[u8]) -> Option<Self> {
        let half = sample.rate < 10000;
        let target_rate = if half { XGM_HALF_RATE } else { XGM_FULL_RATE };

        let resampled = util::resample(
            bank_data,
            sample.addr as usize,
            sample.len as usize,
            sample.rate as usize,
            target_rate as usize,
            256,
        );

        Some(XGMSample::new(id, resampled, half, sample.id, sample.addr))
    }

    pub fn get_length(&self) -> usize {
        self.data.len()
    }

    /// Get similarity score between this sample and another.
    /// Returns 1.0 for exact match (on min-size prefix), 0.0 otherwise.
    /// Advanced fingerprint comparison is not implemented (would require musicg port).
    pub fn get_similarity_score(&self, origin_sample: &XGMSample) -> f64 {
        let delta_size = origin_sample.data.len() as i64 - self.data.len() as i64;
        // origin sample is longer ? cannot use a shorter sample to replace it...
        if delta_size > 150 {
            return 0.0;
        }

        let min_size = self.data.len().min(origin_sample.data.len());

        if crate::sample_advanced_compare() {
            // Advanced comparison not implemented (requires musicg fingerprint library).
            // Fall through to simple comparison.
            if crate::is_verbose() {
                println!(
                    "Warning: advanced sample comparison not available, using simple comparison for #{} and #{}",
                    self.id, origin_sample.id
                );
            }
        }

        // simple sample compare (exact match)
        if self.data[..min_size] == origin_sample.data[..min_size] {
            1.0
        } else {
            0.0
        }
    }
}
